braindecode/preprocessing/preprocess.py

```diff
@@ -120,54 +127,123 @@ def preprocess(concat_ds, preprocessors):
+        else:  # joblib made copies
+            _replace_inplace(concat_ds, BaseConcatDataset(list_of_ds))
 
-    # Recompute cumulative sizes as the transforms might have changed them
-    # XXX: Ultimately, the best solution would be to have cumulative_size be
-    #      a property of BaseConcatDataset.
-    concat_ds.cumulative_sizes = concat_ds.cumsum(concat_ds.datasets)
+        # Store preprocessing keyword arguments in the dataset
+        _set_preproc_kwargs(concat_ds, preprocessors)
+
+    return concat_ds
+
+
+def _replace_inplace(concat_ds, new_concat_ds):
+    """Inplace modification of a BaseConcatDataset with another's attributes.
+
+    Parameters
+    ----------
+    concat_ds : BaseConcatDataset
+        Dataset to modify inplace.
+    new_concat_ds : BaseConcatDataset
+        Dataset to use to modify ``concat_ds``.
+    """
+    if len(concat_ds.datasets) != len(new_concat_ds.datasets):
+        raise ValueError('Both inputs must have the same length.')
+    for i in range(len(new_concat_ds.datasets)):
+        concat_ds.datasets[i] = new_concat_ds.datasets[i]
 
+    concat_kind = 'raw' if hasattr(concat_ds.datasets[0], 'raw') else 'window'
+    preproc_kwargs_attr = concat_kind + '_preproc_kwargs'
+    if hasattr(new_concat_ds, preproc_kwargs_attr):
+        setattr(concat_ds, preproc_kwargs_attr,
+                getattr(new_concat_ds, preproc_kwargs_attr))
 
-def _preprocess(raw_or_epochs, preprocessors):
+
+def _preprocess(ds, ds_index, preprocessors, save_dir=None, overwrite=False):
     """Apply preprocessor(s) to Raw or Epochs object.
 
     Parameters
     ----------
-    raw_or_epochs: mne.io.Raw or mne.Epochs
-        Object to preprocess.
+    ds: BaseDataset | WindowsDataset
+        Dataset object to preprocess.
+    ds_index : int
+        Index of the BaseDataset in its BaseConcatDataset. Ignored if save_dir
+        is None.
     preprocessors: list(Preprocessor)
         List of preprocessors to apply to the dataset.
+    save_dir : str | None
+        If provided, save the preprocessed BaseDataset in the
+        specified directory.
+    overwrite : bool
+        If True, overwrite existing file with the same name.
     """
-    for preproc in preprocessors:
-        preproc.apply(raw_or_epochs)
+
+    def _preprocess_raw_or_epochs(raw_or_epochs, preprocessors):
+        for preproc in preprocessors:
+            preproc.apply(raw_or_epochs)
+
+    if hasattr(ds, 'raw'):
+        _preprocess_raw_or_epochs(ds.raw, preprocessors)
+    elif hasattr(ds, 'windows'):
+        _preprocess_raw_or_epochs(ds.windows, preprocessors)
+    else:
+        raise ValueError(
+            'Can only preprocess concatenation of BaseDataset or '
+            'WindowsDataset, with either a `raw` or `windows` attribute.')
+
+    if save_dir is not None:
+        concat_ds = BaseConcatDataset([ds])
+
+        # Store preprocessing keyword arguments in the dataset
+        _set_preproc_kwargs(concat_ds, preprocessors)
```

Comment:
If this is a BaseConcatDataset of WindowsDatasets, we would also have to set window_kwargs